lib/command.coffee

```diff
@@ -70,12 +70,25 @@ module.exports = class Command
 					return callback?(error) if error?
 
 					try
-						@action(params, parsedOptions, callback)
+						actionPromise = @action(params, parsedOptions, callback)
 					catch error
 						return callback?(error)
 
-					# Means the user is not declaring the callback
-					return callback?() if @action.length < 3
+					if callback?
+						if actionPromise?.then?
+							# Asynchronous action function
+							actionPromise.then(
+								# the value of a fulfilled promise is intentionally ignored because users are expected
+								# to call the callback themselves, and also for backwards compatibility.
+								(_value) => callback() if @action.length < 3  # the user hasn't declared the callback
+
+								# call the callback if the promise is rejected (e.g. if an async action function throws
+								# an error), whether or not the user declared a callback. This is consistent with the
+								# existing implementation for synchronous action functions that throw an error.
+								(_error) -> callback _error if _error)
+						else
+							# else: synchronous action function
+							return callback() if @action.length < 3  # the user hasn't declared the callback
```

Comment:
@jviotti, by the way, the main change here compared to your version in the [Flowdock thread](https://www.flowdock.com/app/rulemotion/r-beginners/threads/cxVr9AuSuVrQ7E53XyshVyBq8BH) is that I had to move `return callback?() if @action.length < 3` from the end of the function to the two if/else branches. Otherwise, in the case of a rejected promise _and_ undeclared callback, `callback()` without arguments would be called before the promise was resolved and then called again once the promise was rejected, this time with an error argument. This was actually causing one of the tests to fail with "[done() called multiple times](https://mochajs.org/#detects-multiple-calls-to-done)" -- so the tests were worth it and "thanks Mocha!" :-)

Reply:
Haha awesome!

Reply:
@zvin, you mean that the CoffeeScript compiler generates the same JavaScript code for following two alternative lines, right?
```
return callback() if @action.length < 3
```
```
callback() if @action.length < 3
```
Yes, if I need to change the code again, I can delete the `return` keyword.